lib/Parse/ParseType.cpp

```diff
@@ -219,23 +258,23 @@ ParserResult<TypeRepr> Parser::parseTypeSimple(
     return nullptr;
   }
 
-  // '.Type', '.Protocol', '?', '!', and '[]' still leave us with type-simple.
+  // '.X', '.Type', '.Protocol', '?', '!', '[]'.
   while (ty.isNonNull()) {
-    if ((Tok.is(tok::period) || Tok.is(tok::period_prefix))) {
-      if (peekToken().isContextualKeyword("Type")) {
-        consumeToken();
-        SourceLoc metatypeLoc = consumeToken(tok::identifier);
-        ty = makeParserResult(ty,
-          new (Context) MetatypeTypeRepr(ty.get(), metatypeLoc));
-        continue;
-      }
-      if (peekToken().isContextualKeyword("Protocol")) {
+    if (Tok.isAny(tok::period, tok::period_prefix)) {
+      if (peekToken().is(tok::code_complete)) {
```

Comment:
I inlined it because I considered it to be a clarity win for all the code completion handling to occur in `parseTypeSimple`, rather than having it scattered across different functions. I’m happy to tuck it back in if you don’t share my vision, though.

Reply:
I understand the rationale. But it wasn't clear for me how `Foo.Bar.<code completion>` would be handled _when seeing `parseTypeDotted()` code_. Maybe modifying `parseTypeDotted()` like the following clarify it.

```diff
-  while (true) {
-     if (Tok.isNot(tok::period, tok::period_prefix) ||
-         peekToken().is(tok::code_complete)) {
-       break;
-     }
-
+  while (Tok.isAny(tok::period, tok::period_prefix)) {
+
+    // '.<code completion>' is handled in `parseTypeSimple()`.
+    if (peekToken().is(tok::code_complete))
+      break
+.  
```

Reply:
Sounds good!